src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,226 @@
+        }
+
+        /// <summary>
+        /// Get the user-specified naming rules, then add standard default naming rules 
+        /// for both static and non-static fields.  The standard naming rules are added at the end 
+        /// so they will only be used if the user hasn't specified a preference.
+        /// </summary>
+        private static async Task<ImmutableArray<NamingRule>> GetNamingRulesAsync(
+            Document document,
+            CancellationToken cancellationToken)
+        {
+            const string defaultStaticFieldPrefix = "s_";
+            const string defaultFieldPrefix = "_";
+
+            var optionSet = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var namingPreferencesOption = optionSet.GetOption(SimplificationOptions.NamingPreferences);
+            var rules = namingPreferencesOption.CreateRules().NamingRules
+                .AddRange(CreateNewRule(ImmutableArray.Create(new ModifierKind(ModifierKindEnum.IsStatic)), defaultStaticFieldPrefix))
+                .AddRange(CreateNewRule(ImmutableArray.Create<ModifierKind>(), defaultFieldPrefix));
+            return rules;
+        }
+
+        private static ImmutableArray<NamingRule> CreateNewRule(
+            ImmutableArray<ModifierKind> modifiers,
+            string prefix)
+        {
+            return ImmutableArray.Create(
+                new NamingRule(
+                    new SymbolSpecification(
+                        Guid.NewGuid(),
+                        "Field",
+                        ImmutableArray.Create(new SymbolSpecification.SymbolKindOrTypeKind(SymbolKind.Field)),
+                        modifiers: modifiers),
+                    new NamingStyles.NamingStyle(
+                        Guid.NewGuid(),
+                        prefix: prefix,
+                        capitalizationScheme: Capitalization.CamelCase),
+                    DiagnosticSeverity.Hidden));
+        }
+
+        private string GenerateFieldName(IPropertySymbol property, ImmutableArray<NamingRule> rules)
+        {
+            var propertyName = property.Name;
+            var fieldName = "";
+            foreach (var rule in rules)
+            {
+                if (rule.SymbolSpecification.AppliesTo(
+                    new SymbolKindOrTypeKind(SymbolKind.Field),
+                    property.IsStatic ? DeclarationModifiers.Static : DeclarationModifiers.None,
+                    Accessibility.Private))
+                {
+                    fieldName = rule.NamingStyle.MakeCompliant(propertyName).First();
+                    break;
+                }
+            }
+
+            return NameGenerator.GenerateUniqueName(fieldName, n => !property.ContainingType.GetMembers(n).Any());
+        }
+
+        internal override (SyntaxNode newGetAccessor, SyntaxNode newSetAccessor) GetNewAccessors(
+            DocumentOptionSet options, SyntaxNode property, 
```

Comment:
Consider making the based typed parameterized on things like TPropertyDeclaration.  That way you don't have to do the casts below.